src/main/java/com/pingcap/tikv/region/RegionManager.java

```diff
@@ -111,31 +117,37 @@ private synchronized TiRegion getRegionById(long regionId) {
      */
     public synchronized void invalidateRegion(long regionId) {
       try {
-        TiRegion region = regionCache.getIfPresent(regionId);
+        if (logger.isDebugEnabled()) {
+          logger.debug(String.format("invalidateRegion ID[%s]", regionId));
+        }
+        TiRegion region = regionCache.get(regionId);
         keyToRegionIdCache.remove(makeRange(region.getStartKey(), region.getEndKey()));
       } catch (Exception ignore) {
       } finally {
-        regionCache.invalidate(regionId);
+        regionCache.remove(regionId);
       }
     }
 
     public synchronized void invalidateAllRegionForStore(long storeId) {
-      for (TiRegion r : regionCache.asMap().values()) {
+      for (TiRegion r : regionCache.values()) {
         if(r.getLeader().getStoreId() == storeId) {
-          regionCache.invalidate(r.getId());
+          if (logger.isDebugEnabled()) {
+            logger.debug(String.format("invalidateAllRegionForStore Region[%s]", r));
+          }
+          regionCache.remove(r.getId());
           keyToRegionIdCache.remove(makeRange(r.getStartKey(), r.getEndKey()));
         }
       }
     }
 
-    public void invalidateStore(long storeId) {
-      storeCache.invalidate(storeId);
+    public synchronized void invalidateStore(long storeId) {
+      storeCache.remove(storeId);
```

Comment:
why change to `remove`?

Reply:
because we use hashmap instead now. No invalidate method anymore.
Previously we use loadingcache since we want to keep low memory. But it seems PD will not have too large number of region and cache all is ok.

Reply:
Gotcha.